src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Async.cs

```diff
@@ -40,5 +45,226 @@ protected override bool CompareValueToValue(AsyncMethodVariant value1, AsyncMeth
             protected override AsyncMethodVariant CreateValueFromKey(EcmaMethod key) => new AsyncMethodVariant(key);
         }
         private AsyncVariantImplHashtable _asyncVariantImplHashtable = new AsyncVariantImplHashtable();
+
+        public MetadataType GetContinuationType(GCPointerMap pointerMap)
+        {
+            return _continuationTypeHashtable.GetOrCreateValue(pointerMap);
+        }
+
+        private sealed class ContinuationTypeHashtable : LockFreeReaderHashtable<GCPointerMap, ContinuationType>
+        {
+            private readonly CompilerTypeSystemContext _parent;
+            private MetadataType _continuationType;
+
+            public ContinuationTypeHashtable(CompilerTypeSystemContext parent)
+                => _parent = parent;
+
+            protected override int GetKeyHashCode(GCPointerMap key) => key.GetHashCode();
+            protected override int GetValueHashCode(ContinuationType value) => value.PointerMap.GetHashCode();
+            protected override bool CompareKeyToValue(GCPointerMap key, ContinuationType value) => key.Equals(value.PointerMap);
+            protected override bool CompareValueToValue(ContinuationType value1, ContinuationType value2)
+                => value1.PointerMap.Equals(value2.PointerMap);
+            protected override ContinuationType CreateValueFromKey(GCPointerMap key)
+            {
+                if (_continuationType == null)
+                    _continuationType = _parent.SystemModule.GetKnownType("System.Runtime.CompilerServices"u8, "Continuation"u8);
+                return new ContinuationType(_continuationType, key);
+            }
+        }
+        private ContinuationTypeHashtable _continuationTypeHashtable;
+
+        /// <summary>
+        /// An async continuation type. The code generator will request this to store local state
+        /// through an async suspension/resumption. We only identify these using a <see cref="GCPointerMap"/>
+        /// since that's all the code generator cares about - size of the type, and where the GC pointers are.
+        /// </summary>
+        private sealed class ContinuationType : MetadataType
+        {
+            private readonly MetadataType _continuationType;
+            private FieldDesc[] _fields;
+            public GCPointerMap PointerMap { get; }
+
+            public override DefType[] ExplicitlyImplementedInterfaces => [];
+            public override ReadOnlySpan<byte> Name => Encoding.UTF8.GetBytes(DiagnosticName);
+            public override ReadOnlySpan<byte> Namespace => [];
+
+            // The layout of the type is "sequential-in-spirit", but since there are GC pointers,
```

Comment:
Note that the JIT sorts the layout before handing it off to the runtime, so it is more like Auto where the actual layout work is done by the JIT.

It may make sense to return true from IsAutoLayout to keep the spirit, but it should not really matter. Nothing should be calling these properties.